### ISSUE TITLE
Snippets: Scope by specific file or pattern

### ISSUE THREAD

**Issue by brpaz** [microsoft/vscode#75955]
Global snippets already support being scoped by language and language file snippets are naturally scoped.  I would like to expand that "scoping" logic to specific files by matching the filename, for example.

A good use case for this, let's say I want to create snippets for travis.yml file. Right now what I do is to create the snippet scoped to YAML language. 

The problem is that it will appear for every YAML file which will pollute the autocomplete with not very useful results if you have snippets for multiple kinds of files that have the same extension. (I don't want to see "travis"  related snippets in a "gitlab-ci.yml" file for example) as they are not applicable.

I think this is most critical for configuration files where they have a very specific scope and that you can easily restrict by looking at the filename.

Going further, we could even expand this concept for other kinds of snippets. For example, I can have some project level snippets for building queries and I only want to enable them, in Repository classes and not in Controller classes. You could scope the files it applies using a regex pattern, for example.

I guess this shouldn't be very hard to do and it would vastly improve the suggestions quality of the snippets.

I see adding a new field similar to "scope" called "context" or something where you can specify a regex that will match to the filename.

**Comment by jrieken** [comment]
Valid request, current workaround is to write an extension that does this.

**Comment by brpaz** [comment]
Might give a try to do an extension. I have searched in the documentation and I believe that the CompletionProvider API is the way to do this. 

Something like this  example(https://github.com/microsoft/vscode-extension-samples/blob/master/completions-sample/src/extension.ts).

Getting the current file name,  match some files containing the snippet code and add the completion items programmatically.

Am I on the right path?

Thank you.

**Comment by jrieken** [comment]
Yes, that's the way to go

**Comment by jrieken** [comment]
should be like a document selector, see https://github.com/microsoft/vscode/issues/78681#issuecomment-519440312

**Comment by GustavoASC** [comment]
Nice feature. We have developed **Rech COBOL** extension and it has built-in support for external JSON snippets and allows filtering each snippet by filename regex, as can be seen on extension details.

![image](https://user-images.githubusercontent.com/10179520/70390184-86cc7300-19a7-11ea-9c8c-864f13aa615e.png)

![image](https://user-images.githubusercontent.com/10179520/70390171-63a1c380-19a7-11ea-9be5-8756943ff3a3.png)

But of course at this moment it's related to this specific COBOL extension. This feature could be refactored and separated into a different extension.

**Comment by kevinkhill** [comment]
I just ran into wanting to do this! Any progress?

I have `jest` snippets and would like to scope them to only suggest in `*.test.*` files :thinking:

**Comment by brpaz** [comment]
@kevinkhill In the meanwhile, I built an [extension](https://github.com/brpaz/vscode-contextual-snips) for this. You might want to check it.

**Comment by kevinkhill** [comment]
Awesome! I will give it a spin next ime I get on my laptop.

**Comment by cesarcf** [comment]
Hi @brpaz,  Install the extension and I have created a snippet with the create command but it seems that it is not working ...
is the extension working correctly?
`{
  "test description": {
    "prefix": ["test"],
    "body": ["singleQuote: false"],
    "context": {
      "patterns": ["**/**/prettier.config.js"]
    },
    "id": "test description",
    "scope": "javascript",
    "description": "test description"
  }
}`


Is there any other alternative to do this?

**Comment by brpaz** [comment]
@cesarcf Hi. Yes, I am using my extension frequently. It should work. Feel free to open a dedicated issue in my repository.

**Comment by zardoy** [comment]
Hey @brpaz, I'm definitely building a very similar extension [Better Snippets](https://marketplace.visualstudio.com/items?itemName=zardoy.better-snippets) but with most configurable context settings I could ever imagine.

Just like yours extension it supports:

- restricting by absolute path. But it's using a single regex instead of array of globs, so `**/extension.ts` would be equivalent to `extension.ts` (strict variant: `\\/extension.ts$`). Also it comes with shortcuts for complex regex such as tsconfig – `when.fileType`.
- restricting by line contents:
  - `when.lineHasRegex`: regex against whole line
  - `when.lineRegex`: regex against line till cursor position
- the fastest way to create snippets via `Create Snippet from Selection`

But also:

- - captured named groups from file or line regex can be used within snippet template (body). For example:

Snippet config:

```json
{
    "name": "createAsyncThunk",
    "body": "createAsyncThunk('$NAME', async ($1) => $2)",
    "sortText": "!",
    "when": {
        "pathRegex": "\\/(?<NAME>[^\\/]+)\\/slice\\.ts$",
        "locations": ["topLineStart"]
    }
},
```

Using it in `/someProjectPath/src/redux/users/slice.ts` will expand to `createAsyncThunk('users', async () => )`. Obviously in other files (such as `index.ts`) the snippet won't be displayed.


- restricting by cursor position: `when.locations`
This one I use the most as there are snippets that you use only on empty lines: [useState](https://github.com/zardoy/vscode-better-snippets/blob/58ffbeee1edda4e0061275f6639fb3ab3b24fdf0/test/snippets.jsonc#L4-L22) for example.
- resolve snippet imports (thats what I'm missing most from every IDE)
- configure sortText and icon via `iconType` (completely custom icon will be available after CustomUI patch)
- optionally set custom file or folder icon (useful with file icon themes)

But I like that you allow restrict snippets by package (though it'd good to support monorepos).

Also IMO having snippets defined in `settings.json` (just like [TS/JS Postfix](https://marketplace.visualstudio.com/items?itemName=ipatalas.vscode-postfix-ts) and Surround do) is more reliable and transparent for the end user.

And also, every feature of this extensions works on the web. Would like to hear any thougts!

**Comment by brpaz** [comment]
@zardoy hey. Love it. I will give it a try.

**Comment by erayerdin** [comment]
My usecase is this: I have a Flutter project. I want to insert snippets that creates tests and test groups. However, I want these snippets to appear only in files with the names of which end with `_test.dart`. It shouldn't appear in other files.

**Comment by rudyflores** [comment]
Any updates on this issue @jrieken ? I am looking into being able to filter snippets for specific filenames as well in my project.

**Comment by zardoy** [comment]
> extension is very cool and all but the settings of a custom extension does not sync AFAIR

@erayerdin No, why? All settings live in `settings.json` which of course is synced (until you explicity disable it for some reason). It would be hard to live for me without custom extension snippets sync.

**Comment by shijunti19** [comment]
@jrieken Code prompt is too slow，It should be optimized and accelerated. It is better to cache

**Comment by equiman** [comment]
> I have `jest` snippets and would like to scope them to only suggest in `*.test.*` files 🤔

I'm trying the same with some `vitest` snippets. I've tried adding this:

```
{
  "contributes": {
    "languages": [
      {
        "id": "vue",
        "aliases": [
          "Vue",
          "vue"
        ],
        "extensions": [
          ".test.vue",
          ".spec.vue"
        ]
      }
    ]
  }
}
```

but doesn't work because is still showing the snippets on each vue file.

**Comment by doggy8088** [comment]
GitHub Copilot seems give us a good suggestion. 😄

![image](https://github.com/microsoft/vscode/assets/88981/f5dfc1a1-8aa4-4692-aece-798a56711203)

**Comment by equiman** [comment]
Well, I've tried with:

``` 
"when": {
  "pathRegex": "\\/(?<NAME>[^\\/]+)\\/.(test|spec)\\.(js|jsx|ts|tsx)$"
}
```

```
"when": "editorFileName =~ /\\.(test|spec)\\.(js|jsx|ts|tsx)$/"
```


```
"when": "resourceFileName =~ /\\.(test|spec)\\.(js|jsx|ts|tsx)$/"
```

```
"scope": [
      "*.test.js",
      "*.test.jsx",
      "*.test.ts",
      "*.test.tsx",
      "*.spec.js",
      "*.spec.jsx",
      "*.spec.ts",
      "*.spec.tsx"
    ]
```

No one stops to show the snippets on all files.
